R/snap_tools.R

```diff
@@ -0,0 +1,512 @@
+      
+      
+      # Find vertices coordinates corresponding
+      # to the bulk of the original segment which
+      # aren't those between the elongated vertex
+      # and the junction vertex with the bridge
+      coords_splitted <- sf::st_coordinates(road_features)
+      idx_duplicated <- which((diff(coords_splitted[,"X"]) == 0) & (diff(coords_splitted[,"Y"]) == 0))
+      coords_splitted <- coords_splitted[-idx_duplicated,]
+      
+      x_match_elong <- coords_splitted[,"X"] == x_elongated
+      y_match_elong <- coords_splitted[,"Y"] == y_elongated
+      idx_match_elong <- which(x_match_elong & y_match_elong)
+      
+      # Some wiggle room must be added due to slight imprecision
+      junction_point <- dist_bridge[idx_valid_intersection[j]] |>
+        st_point_on_path(bridge)
+      x_match_junc <- (coords_splitted[,"X"] >= (junction_point["X"]-0.01)) &
+                      (coords_splitted[,"X"] <= (junction_point["X"]+0.01))
+      y_match_junc <- (coords_splitted[,"Y"] >= (junction_point["Y"]-0.01)) &
+                      (coords_splitted[,"Y"] <= (junction_point["Y"]+0.01))
+      idx_match_junc <- which(x_match_junc & y_match_junc)
+      
+      # Remove all coordinates between the elongated vertex
+      # and the junction vertex
+      coords_keep <- coords_splitted[-(idx_match_junc:idx_match_elong),-3]
+      
+      # Edit the coordinates at (mean) splitting point
+      idx_junction <- if(n != 1) nrow(coords_keep) else 1
+      coords_keep[idx_junction,] <- coords_junction
+      
+      # Update geometry of the corrected segment
+      sf::st_geometry(roads[pos,]) <- coords_keep |>
+        sf::st_linestring() |>
+        sf::st_sfc(crs = sf::st_crs(roads))
+      
+      
+      # Update "limit" values in tb_ends_roads table
+      # as the splitting operation might change the 
+      # total number of vertices in the geometry
+      tb_ends_roads[pos, "limit"] <- nrow(coords_keep)
+    }
+  }
+  
+  return(roads)
+}
+
+
+bridge_intersection <- function(j, roads, bridge, bridge_junction, tb_node_remain)
+{
+  pos <- tb_node_remain[j,][["pos"]]
+  n <- tb_node_remain[j,][["limit"]]
+  heading <- tb_node_remain[j,][["heading"]]*pi/180
+  
+  road_copy <- roads[pos,] # Copie peut-être pas nécessaire
+  
+  x_remain <- sf::st_geometry(road_copy)[[1]][n, 1] + 999999 * cos(heading)
+  y_remain <- sf::st_geometry(road_copy)[[1]][n, 2] + 999999 * sin(heading)
+  
+  sf::st_geometry(road_copy)[[1]][n, 1] <- x_remain
+  sf::st_geometry(road_copy)[[1]][n, 2] <- y_remain
+  
+  pts_intersect <- sf::st_intersection(bridge, sf::st_geometry(road_copy))
+  
+  if (length(pts_intersect) == 0)
+  {
+    x_old <- round(sf::st_geometry(roads[pos,])[[1]][n,1])
+    y_old <- round(sf::st_geometry(roads[pos,])[[1]][n,2])
+    warning(glue::glue("{tb_node_remain[j,][['id']]} never crosses bridge. This end of the road won't be snapped. X:{x_old} Y:{y_old}"), call.=FALSE)
```

Comment:
Cette erreur n'est pas comprhensibe quand on a pas codé le truc et qu'on a pas lu le code source our comprendre ce qu'est un "bridge" en plus ca donne un ID genre `1234567_end`. J'i'magine qu'on voudrais plutot un warning genre

> At end of the road 1234567: no spapping because ....

Reply:
J'ai décidé de retirer l'avertissement car il est redondant avec celui à L232 qui fait le même travail au final.